cpp/src/arima/batched_kalman.cu

```diff
@@ -659,17 +666,67 @@ __global__ void confidence_intervals(const double* d_fc, const double* d_sigma2,
   d_upper[idx] = fc + margin;
 }
 
+void _lyapunov_wrapper(raft::handle_t& handle,
+                       const ARIMAMemory<double>& arima_mem,
+                       const MLCommon::LinAlg::Batched::Matrix<double>& A,
+                       MLCommon::LinAlg::Batched::Matrix<double>& Q,
+                       MLCommon::LinAlg::Batched::Matrix<double>& X, int r) {
+  if (r <= 5) {
+    auto stream = handle.get_stream();
+    auto cublasHandle = handle.get_cublas_handle();
+    auto allocator = handle.get_device_allocator();
+    int batch_size = A.batches();
+    int r2 = r * r;
+    auto counting = thrust::make_counting_iterator(0);
+
+    //
+    // Use direct solution with Kronecker product
+    //
+
+    MLCommon::LinAlg::Batched::Matrix<double> I_m_AxA(
+      r2, r2, batch_size, cublasHandle, arima_mem.I_m_AxA_batches,
+      arima_mem.I_m_AxA_dense, allocator, stream, false);
+    MLCommon::LinAlg::Batched::Matrix<double> I_m_AxA_inv(
+      r2, r2, batch_size, cublasHandle, arima_mem.I_m_AxA_inv_batches,
+      arima_mem.I_m_AxA_inv_dense, allocator, stream, false);
+
+    MLCommon::LinAlg::Batched::b_kron(A, A, I_m_AxA, -1.0);
+
+    double* d_I_m_AxA = arima_mem.I_m_AxA_dense;
+    thrust::for_each(thrust::cuda::par.on(stream), counting,
+                     counting + batch_size, [=] __device__(int ib) {
+                       double* b_I_m_AxA = d_I_m_AxA + ib * r2 * r2;
+                       for (int i = 0; i < r2; i++) {
+                         b_I_m_AxA[(r2 + 1) * i] += 1.0;
+                       }
+                     });
+
+    MLCommon::LinAlg::Batched::Matrix<double>::inv(
+      I_m_AxA, I_m_AxA_inv, arima_mem.I_m_AxA_P, arima_mem.I_m_AxA_info);
+
+    Q.reshape(r2, 1);
+    X.reshape(r2, 1);
+    MLCommon::LinAlg::Batched::b_gemm(false, false, r2, 1, r2, 1.0, I_m_AxA_inv,
+                                      Q, 0.0, X);
+    Q.reshape(r, r);
+    X.reshape(r, r);
```

Comment:
Can we decrease code duplication between here and `b_lyapunov` by placing some of these into a helper function?